Add a helper method to judge whether a given instance has been enhanced

ShaneXiangH opened the pull request:
Solve issue #187 

I add a marker interface in the file bytecodeGen.java 

Each time a new Enhanced class is generated, it also generate a instance with marker interface.

mcculls commented:
Marker interfaces don't play well with modular systems like [JPMS](https://www.jcp.org/en/jsr/detail?id=376) because the enhanced class needs to be able to access both the type it's enhancing and the marker interface, which is likely defined in a different classloader. This means you need to use a "bridge" classloader for the enhanced class which introduces a lot of extra complexity.

The new enhancer code in #1298 deliberately avoids introducing Guice-specific types in the enhanced class, which keeps the classloading very simple. Introducing a Guice-specific marker interface here would undo this work and mean we'd need to bring back bridge class-loading.

I'd also question what this marker interface would be used for - ie. why do clients need to know if the type has been enhanced or not, and why can't they look at the class name or hierarchy (Guice uses a naming convention for enhanced types which makes this relatively straightforward.)

An alternative solution would be to introduce a static helper that returned whether a given class or instance has been enhanced. This could use the naming convention as described above to decide and would avoid the need to introduce a dependency back to Guice from the enhanced type.

ShaneXiangH commented:
mcculls thank you. I think you are right.

I added such a static helper method in Guice.java and this method can returns whether the given instance has been enhanced. In addition, I added a testing method in BytecodeGenTest.java

alex-courtis commented:
`Guice#isGuiceEnhanced` successfully passes `#testIsEnhancedJudge` on java 1.8 maven 3.2.5.

I wonder if we might add `stripGuiceMarksFromClassName` to this PR, as per https://github.com/google/guice/issues/187#issuecomment-327768611
There is a real world use case for it and the test is almost the same.
